### ISSUE TITLE
Problems related to mixing static and dynamic linking

### ISSUE THREAD

**Issue by fedetft** [OpenModelica/OpenModelica#8955]
As suggested by @casella, starting from the discussion in  #8738, I'm opening a ticket about issues that are popping up having as root cause the mixing of static and dynamic linking.

My way to see the problem, is to imagine a process in memory as a graph, where the root node is the executable, and all other nodes are libraries. Each arc in the graph is a linking operation, that can either be static or dynamic.
Problems potentially arise when multiple nodes in the graph need to link with the same library, as to avoid undefined behavior some means of deduplication is mandatory to make sure there is one and only one instance of the global variables and functions of a library, that pointers to the same functions have the same value, as well as to avoid the bug of linking with different versions of the same library that are subtly incompatible.

Trouble is, all this is done automatically by the dynamic linker ONLY if everything is linked dynamically.

As soon as you link something statically, YOU become responsible to be sure that the library you're linking statically is linked only once in that graph.
In #8738, at least two instances where this constraint was violated, with `libstdc++` and `libomcgc` being linked statically at least once, and both resulting in segfaults.

The previous ticket suggested that the mixing of static and dynamic linking is caused by the scale of the problem, with @mahge suggesting that openmodelica is composed of 100 targets, but I argue otherwise. Static linking is unmanageable exactly because of the scale of the problem, as I don't see you're going to draw that graph by hand and carefully check no duplicate linking with at least one static arc occur in it, let alone keep that graph updated every time you make a commit.

To be clear, I don't have a strong solution to the issue, linking everything dynamically works but there may be reasons that could make it impractical, and I'd be glad to know them. I'm just pointing out that the situation is or may soon become unmanageable.
And now let the discussion begin ;)

**Comment by mahge** [comment]
>  there may be reasons that could make it impractical, and I'd be glad to know them.

Spoken like a true linux user :). There is Windows and there is the lack of `rpath` which I think is the reason why most of this static stuff exists. To avoid having DLLs and having to move them around for each simulation executable. Luckily that has been practically solved with #8572 and follow up fixes in a way that can work well with OMEdit. 

> To be clear, I don't have a strong solution to the issue,

Luckily there is an ideal solution. It is to use a build system generator like CMake. For many more reasons than just this. Even better news is that we already have it for OpenModelica. It is just not the default build system yet. It needs to be tested by more people and polished a bit more. The Windows segfault does not happen with the CMake built libraries and omc.exe even though it builds those exact libraries (`libOpenModelicaRuntimeC` and `libomcgc `) as static as well, just to resemble the default build's outputs.

**Comment by mahge** [comment]
> There is Windows and there is the lack of rpath which I think is the reason why most of this static stuff exists

I forgot that we have so many 3rdParty libraries that we build as well. Most of them are static. We can attempt to change them but that is a monumental undertaking.

**Comment by fedetft** [comment]
> > To be clear, I don't have a strong solution to the issue,
> 
> Luckily there is an ideal solution. It is to use a build system generator like CMake. For many more reasons than just this. Even better news is that we already have it for OpenModelica. It is just not the default build system yet. It needs to be tested by more people and polished a bit more. The Windows segfault does not happen with the CMake built libraries and omc.exe even though it builds those exact libraries (`libOpenModelicaRuntimeC` and `libomcgc `) as static as well, just to resemble the default build's outputs.

CMake is indeed wonderful and if you're in the process of switching to it, by all means do. You won't regret it.

But it won't deduplicate at runtime global variables of statically linked libraries, so the root cause of those segfaults will still be there lurking.

**Comment by mahge** [comment]
> But it won't deduplicate at runtime global variables of statically linked libraries, so the root cause of those segfaults will still be there lurking.

Yes in this case we have now gone with building the libraries as shared since that is the right thing to do and there seems to be no reason to do so otherwise. 

In cases where some libraries have to be static then we can rely on the build system generator to minimize duplicate instances and hope that people avoid global variables 🤞

**Comment by fedetft** [comment]
As I said in the other bug report, thanks a lot for fixing the immediate issue with externalmedia.
If I were you, I'd also try to avoid linking `libstdc++` statically to prevent future issues.
If you want to keep linking `antlr` statically, that doesn't seem too bad to me, as it does not seem a likely library someone would want to use in an external function or in an FMU.

**Comment by mahge** [comment]
> If I were you, I'd also try to avoid linking libstdc++ statically to prevent future issues.

Absolutely. That is next. I could not put it with these fixes since if something goes wrong it will be difficult to hunt down. It is better to separate them. I am going to try and remove the static linking and the manual `libstdc+`+ linkages as well.
